module/zfs/arc.c

```diff
@@ -4658,30 +4664,46 @@ static uint64_t
 arc_all_memory(void)
 {
 #ifdef _KERNEL
-	return (MIN(ptob(physmem),
-	    vmem_size(heap_arena, VMEM_FREE | VMEM_ALLOC)));
+#ifdef CONFIG_HIGHMEM
+	return (ptob(totalram_pages - totalhigh_pages));
+#else
+	return (ptob(totalram_pages));
+#endif
 #else
 	return (ptob(physmem) / 2);
 #endif
 }
 
-#ifdef _KERNEL
+/*
+ * Return the amount of memory that is considered free.  In user space
+ * which is primarily used for testing we pretend that free memory ranges
+ * from 0-20% of all memory.
+ */
 static uint64_t
 arc_free_memory(void)
 {
+#ifdef _KERNEL
+#ifdef CONFIG_HIGHMEM
+	struct sysinfo si;
+	si_meminfo(&si);
+	return (ptob(si.freeram - si.freehigh));
+#else
 #ifdef ZFS_GLOBAL_NODE_PAGE_STATE
-	return (nr_free_pages() +
+	return (ptob(nr_free_pages() +
 	    global_node_page_state(NR_INACTIVE_FILE) +
 	    global_node_page_state(NR_INACTIVE_ANON) +
-	    global_node_page_state(NR_SLAB_RECLAIMABLE));
+	    global_node_page_state(NR_SLAB_RECLAIMABLE)));
 #else
-	return (nr_free_pages() +
+	return (ptob(nr_free_pages() +
 	    global_page_state(NR_INACTIVE_FILE) +
 	    global_page_state(NR_INACTIVE_ANON) +
-	    global_page_state(NR_SLAB_RECLAIMABLE));
+	    global_page_state(NR_SLAB_RECLAIMABLE)));
+#endif /* ZFS_GLOBAL_NODE_PAGE_STATE */
+#endif /* CONFIG_HIGHMEM */
+#else
+	return (spa_get_random(arc_all_memory() * 20 / 100));
 #endif
```

Comment:
Maybe add a /* KERNEL */ here just to make it clearer to read. I had a hard time figuring out what the spa_get_random was for.

Reply:
Will do